File: packages/neptune_flutter_ui/lib/src/widgets/neptune_wallet_pay.dart
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0

import 'package:flutter/material.dart';

import '../theme/extensions.dart';
import '../theme/neptune_theme.dart';

/// A transaction-like merchant row (web `<npt-merchant-row>`): a circular
/// merchant logo/avatar (or the name's initial), the merchant name + category,
/// and a trailing tabular amount + time. When [pending] the amount dims and a
/// "pending" pill appears. Theme-only, RTL-safe.
class NeptuneMerchantRow extends StatelessWidget {
  final String name;
  final String? category;
  final String amount;
  final String? time;
  final bool pending;
  final Widget? logo;

  const NeptuneMerchantRow({
    super.key,
    required this.name,
    this.category,
    required this.amount,
    this.time,
    this.pending = false,
    this.logo,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final type = Theme.of(context).extension<NptType>()!;
    final text = Theme.of(context).textTheme;
    final amountColor = pending ? scheme.onSurfaceVariant : scheme.onSurface;
    final money = NeptuneTheme.moneyStyle(context, base: text.bodyLarge)
        .copyWith(color: amountColor, fontWeight: FontWeight.w600);
    final initial =
        name.trim().isEmpty ? '•' : name.trim().substring(0, 1).toUpperCase();

    return Container(
      constraints: const BoxConstraints(minHeight: 56),
      decoration: BoxDecoration(
        border: Border(bottom: BorderSide(color: scheme.outlineVariant)),
      ),
      padding: const EdgeInsetsDirectional.symmetric(horizontal: 8, vertical: 12),
      child: Row(
        children: [
          // Logo / avatar.
          ClipOval(
            child: Container(
              width: 44,
              height: 44,
              color: scheme.secondaryContainer,
              alignment: Alignment.center,
              child: logo ??
                  Text(
                    initial,
                    style: text.labelLarge?.copyWith(
                      fontFamily: type.display,
                      fontWeight: FontWeight.w700,
                      color: scheme.onSecondaryContainer,
                    ),
                  ),
            ),
          ),
          const SizedBox(width: 16),
          // Name + category/pending.
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(
                  name,
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: text.bodyLarge?.copyWith(color: scheme.onSurface),
                ),
                if (category != null || pending) ...[
                  const SizedBox(height: 2),
                  Row(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      if (category != null)
                        Flexible(
                          child: Text(
                            category!,
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                            style: text.bodyMedium
                                ?.copyWith(color: scheme.onSurfaceVariant),
                          ),
                        ),
                      if (pending) ...[
                        if (category != null) const SizedBox(width: 8),
                        Container(
                          padding: const EdgeInsetsDirectional.symmetric(
                              horizontal: 8, vertical: 1),
                          decoration: BoxDecoration(
                            color: scheme.tertiaryContainer,
                            borderRadius: BorderRadius.circular(shape.full),
                          ),
                          child: Text(
                            'pending',
                            style: text.labelSmall?.copyWith(
                              color: scheme.onTertiaryContainer,
                              fontWeight: FontWeight.w600,
                            ),
                          ),
                        ),
                      ],
                    ],
                  ),
                ],
              ],
            ),
          ),
          const SizedBox(width: 16),
          // Trailing amount + time.
          Column(
            crossAxisAlignment: CrossAxisAlignment.end,
            mainAxisSize: MainAxisSize.min,
            children: [
              Text(amount, style: money),
              if (time != null) ...[
                const SizedBox(height: 2),
                Text(
                  time!,
                  style: NeptuneTheme.moneyStyle(context, base: text.labelSmall)
                      .copyWith(color: scheme.onSurfaceVariant),
                ),
              ],
            ],
          ),
        ],
      ),
    );
  }
}

/// A reward/voucher coupon (web `<npt-voucher-card>`): a big tabular value stub
/// separated by a dashed tear line from the title, an optional dashed/mono code
/// chip, and an expiry caption. Painted on the primary container. Theme-only,
/// RTL-safe.
class NeptuneVoucherCard extends StatelessWidget {
  final String title;
  final String value;
  final String? code;
  final String? expiry;

  const NeptuneVoucherCard({
    super.key,
    required this.title,
    required this.value,
    this.code,
    this.expiry,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final type = Theme.of(context).extension<NptType>()!;
    final text = Theme.of(context).textTheme;
    final onContainer = scheme.onPrimaryContainer;
    final valueStyle =
        NeptuneTheme.moneyStyle(context, base: text.headlineSmall).copyWith(
      color: onContainer,
      fontWeight: FontWeight.w700,
    );

    return Container(
      decoration: BoxDecoration(
        color: scheme.primaryContainer,
        borderRadius: shape.rLg,
        boxShadow: [
          BoxShadow(
            color: scheme.shadow.withValues(alpha: 0.2),
            blurRadius: 3,
            offset: const Offset(0, 1),
          ),
        ],
      ),
      padding: const EdgeInsetsDirectional.all(20),
      child: IntrinsicHeight(
        child: Row(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          // Value stub with a dashed tear line on the inline-end edge.
          IntrinsicHeight(
            child: Container(
              padding: const EdgeInsetsDirectional.only(end: 16),
              decoration: BoxDecoration(
                border: BorderDirectional(
                  end: BorderSide(color: scheme.outline, width: 2),
                ),
              ),
              alignment: Alignment.center,
              child: Text(value, style: valueStyle),
            ),
          ),
          const SizedBox(width: 16),
          // Body: title, code chip, expiry.
          Expanded(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.start,
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(
                  title,
                  style: text.titleMedium?.copyWith(
                    fontFamily: type.display,
                    fontWeight: FontWeight.w600,
                    color: onContainer,
                  ),
                ),
                if (code != null || expiry != null) ...[
                  const SizedBox(height: 8),
                  Wrap(
                    spacing: 12,
                    runSpacing: 4,
                    crossAxisAlignment: WrapCrossAlignment.center,
                    children: [
                      if (code != null)
                        Container(
                          padding: const EdgeInsetsDirectional.symmetric(
                              horizontal: 12, vertical: 4),
                          decoration: BoxDecoration(
                            color: scheme.surface,
                            borderRadius: shape.rXs,
                          ),
                          child: Text(
                            code!,
                            style: NeptuneTheme.moneyStyle(context,
                                    base: text.labelLarge)
                                .copyWith(
                              color: scheme.onSurface,
                              fontWeight: FontWeight.w600,
                              letterSpacing: 1.2,
                            ),
                          ),
                        ),
                      if (expiry != null)
                        Text(
                          expiry!,
                          style: text.bodySmall?.copyWith(
                            color: onContainer.withValues(alpha: 0.78),
                          ),
                        ),
                    ],
                  ),
                ],
              ],
            ),
          ),
        ],
      ),
      ),
    );
  }
}

/// A scan-to-pay panel (web `<npt-qr-pay>`): a bordered QR placeholder sized
/// [size], an optional merchant caption, and a big tabular amount. Theme-only,
/// RTL-safe.
class NeptuneQrPay extends StatelessWidget {
  final String amount;
  final String? merchant;
  final double size;

  const NeptuneQrPay({
    super.key,
    required this.amount,
    this.merchant,
    this.size = 180,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;
    final money =
        NeptuneTheme.moneyStyle(context, base: text.displaySmall).copyWith(
      color: scheme.onSurface,
      fontWeight: FontWeight.w700,
    );

    return Container(
      decoration: BoxDecoration(
        color: scheme.surfaceContainerLow,
        borderRadius: shape.rXl,
      ),
      padding: const EdgeInsetsDirectional.all(24),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: [
          // QR placeholder.
          Container(
            width: size,
            height: size,
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: scheme.surface,
              borderRadius: shape.rMd,
              border: Border.all(color: scheme.outlineVariant, width: 2),
            ),
            child: Icon(
              Icons.qr_code_2,
              size: size * 0.7,
              color: scheme.onSurfaceVariant,
            ),
          ),
          if (merchant != null) ...[
            const SizedBox(height: 20),
            Text(
              merchant!,
              textAlign: TextAlign.center,
              style: text.bodyMedium?.copyWith(color: scheme.onSurfaceVariant),
            ),
          ],
          const SizedBox(height: 8),
          Text(amount, style: money),
        ],
      ),
    );
  }
}

/// A selectable top-up option row (web `<npt-topup-row>`): a squared icon tile
/// slot, a label + optional sublabel, a trailing tabular amount, and a chevron
/// that mirrors in RTL. Theme-only, RTL-safe.
class NeptuneTopupRow extends StatelessWidget {
  final String label;
  final String? sublabel;
  final String amount;
  final VoidCallback? onTap;

  const NeptuneTopupRow({
    super.key,
    required this.label,
    this.sublabel,
    required this.amount,
    this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final text = Theme.of(context).textTheme;
    final money = NeptuneTheme.moneyStyle(context, base: text.bodyLarge)
        .copyWith(color: scheme.onSurface, fontWeight: FontWeight.w600);

    return Material(
      color: scheme.surfaceContainerLow,
      borderRadius: shape.rMd,
      clipBehavior: Clip.antiAlias,
      child: InkWell(
        onTap: onTap,
        child: Container(
          constraints: const BoxConstraints(minHeight: 56),
          padding:
              const EdgeInsetsDirectional.symmetric(horizontal: 16, vertical: 12),
          child: Row(
            children: [
              // Icon tile.
              Container(
                width: 40,
                height: 40,
                decoration: BoxDecoration(
                  color: scheme.secondaryContainer,
                  borderRadius: shape.rSm,
                ),
                alignment: Alignment.center,
                child: Icon(
                  Icons.add_card,
                  size: 24,
                  color: scheme.onSecondaryContainer,
                ),
              ),
              const SizedBox(width: 16),
              // Label + sublabel.
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Text(
                      label,
                      maxLines: 1,
                      overflow: TextOverflow.ellipsis,
                      style: text.bodyLarge?.copyWith(color: scheme.onSurface),
                    ),
                    if (sublabel != null) ...[
                      const SizedBox(height: 2),
                      Text(
                        sublabel!,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: text.bodySmall
                            ?.copyWith(color: scheme.onSurfaceVariant),
                      ),
                    ],
                  ],
                ),
              ),
              const SizedBox(width: 12),
              Text(amount, style: money),
              const SizedBox(width: 8),
              Icon(
                Icons.chevron_right,
                size: 24,
                color: scheme.onSurfaceVariant,
                textDirection: Directionality.of(context),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

/// The membership tier tones for [NeptuneTierBadge], mirroring the web
/// `tone` attribute (`gold`/`silver`/`primary`/`neutral`).
enum NeptuneTierTone { neutral, primary, gold, silver }

/// A small membership tier pill (web `<npt-tier-badge>`): a leading dot + the
/// tier label, coloured from the theme by [tone]. Theme-only, RTL-safe.
class NeptuneTierBadge extends StatelessWidget {
  final String tier;
  final String? tone;

  const NeptuneTierBadge({
    super.key,
    required this.tier,
    this.tone,
  });

  NeptuneTierTone get _tone => switch (tone?.toLowerCase()) {
        'primary' => NeptuneTierTone.primary,
        'gold' => NeptuneTierTone.gold,
        'silver' => NeptuneTierTone.silver,
        _ => NeptuneTierTone.neutral,
      };

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final shape = Theme.of(context).extension<NptShape>()!;
    final type = Theme.of(context).extension<NptType>()!;
    final text = Theme.of(context).textTheme;

    final (Color bg, Color fg, Color? border) = switch (_tone) {
      NeptuneTierTone.primary => (
          scheme.primaryContainer,
          scheme.onPrimaryContainer,
          null,
        ),
      NeptuneTierTone.gold => (
          scheme.tertiaryContainer,
          scheme.onTertiaryContainer,
          null,
        ),
      NeptuneTierTone.silver => (
          scheme.secondaryContainer,
          scheme.onSecondaryContainer,
          null,
        ),
      NeptuneTierTone.neutral => (
          scheme.surfaceContainerHighest,
          scheme.onSurface,
          scheme.outlineVariant,
        ),
    };

    return Container(
      constraints: const BoxConstraints(minHeight: 24),
      padding:
          const EdgeInsetsDirectional.symmetric(horizontal: 12, vertical: 2),
      decoration: BoxDecoration(
        color: bg,
        borderRadius: BorderRadius.circular(shape.full),
        border: border == null ? null : Border.all(color: border),
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            width: 6,
            height: 6,
            decoration: BoxDecoration(color: fg, shape: BoxShape.circle),
          ),
          const SizedBox(width: 4),
          Text(
            tier,
            style: text.labelLarge?.copyWith(
              fontFamily: type.display,
              fontWeight: FontWeight.w700,
              color: fg,
              letterSpacing: 0.28,
            ),
          ),
        ],
      ),
    );
  }
}

File: packages/neptune_kmp_ui/odyssey-compose-ui/src/commonMain/kotlin/ly/neptune/odyssey/ui/components/NeptuneWalletPay.kt
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// Wallet & pay surfaces. Web counterparts: `<npt-merchant-row>`,
// `<npt-voucher-card>`, `<npt-qr-pay>`, `<npt-topup-row>`, `<npt-tier-badge>`
// (wallet-pay.ts) · Flutter: neptune_wallet_pay.dart. Money renders in the
// brand money style (tabular figures, numerals lever applied). Theme-only,
// RTL-safe, rows ellipsize at narrow widths.

package ly.neptune.odyssey.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.hapticfeedback.HapticFeedback
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import ly.neptune.odyssey.ui.glyphs.NptWalletPayGlyphs
import ly.neptune.odyssey.ui.identity.nptShadow
import ly.neptune.odyssey.ui.theme.NeptuneTheme
import ly.neptune.odyssey.ui.theme.NptFeedbackCue
import ly.neptune.odyssey.ui.theme.rememberNeptuneFontFamily

// The tear/dashed-border pattern shared with the Dart _DashedBorderPainter
// (neptune_card_controls.dart): 6 on, 5 off.
private const val DASH_ON = 6f
private const val DASH_OFF = 5f

/**
 * A transaction-like merchant row: a circular merchant logo/avatar (or the
 * name's initial), the merchant name + category, and a trailing tabular
 * amount + time. When [pending] the amount dims to `onSurfaceVariant` and a
 * "pending" pill appears next to the category.
 *
 * Web counterpart: `<npt-merchant-row>` · Flutter: `NeptuneMerchantRow`.
 *
 * [logo] fills the 44dp circular `secondaryContainer` avatar; when null the
 * name's first initial renders in the display face. Theme-only, RTL-safe.
 */
@Composable
public fun NeptuneMerchantRow(
    name: String,
    amount: String,
    modifier: Modifier = Modifier,
    category: String? = null,
    time: String? = null,
    pending: Boolean = false,
    logo: (@Composable () -> Unit)? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val display = rememberNeptuneFontFamily(NeptuneTheme.type.display)

    val amountColor = if (pending) scheme.onSurfaceVariant else scheme.onSurface
    val money = NeptuneTheme.moneyStyle(base = typography.bodyLarge)
        .copy(color = amountColor, fontWeight = FontWeight.W600)
    val timeStyle = NeptuneTheme.moneyStyle(base = typography.labelSmall)
        .copy(color = scheme.onSurfaceVariant)

    val trimmed = name.trim()
    val initial = if (trimmed.isEmpty()) "•" else trimmed.take(1).uppercase()
    val divider = scheme.outlineVariant

    Row(
        modifier = modifier
            // The 1dp bottom hairline (web border-bottom outline-variant).
            .drawBehind {
                val stroke = 1.dp.toPx()
                drawLine(
                    color = divider,
                    start = Offset(0f, size.height - stroke / 2),
                    end = Offset(size.width, size.height - stroke / 2),
                    strokeWidth = stroke,
                )
            }
            .defaultMinSize(minHeight = 56.dp)
            .padding(horizontal = 8.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // Logo / avatar (falls back to the name's initial).
        Box(
            modifier = Modifier
                .size(44.dp)
                .clip(NeptuneTheme.shape.rFull)
                .background(scheme.secondaryContainer),
            contentAlignment = Alignment.Center,
        ) {
            CompositionLocalProvider(LocalContentColor provides scheme.onSecondaryContainer) {
                if (logo != null) {
                    logo()
                } else {
                    Text(
                        initial,
                        style = typography.labelLarge.copy(
                            fontFamily = display ?: typography.labelLarge.fontFamily,
                            fontWeight = FontWeight.W700,
                            color = scheme.onSecondaryContainer,
                        ),
                    )
                }
            }
        }
        Spacer(Modifier.width(16.dp))
        // Name + category/pending.
        Column(Modifier.weight(1f)) {
            Text(
                text = name,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                style = typography.bodyLarge.copy(color = scheme.onSurface),
            )
            if (category != null || pending) {
                Spacer(Modifier.height(2.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    if (category != null) {
                        Text(
                            text = category,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                            style = typography.bodyMedium.copy(color = scheme.onSurfaceVariant),
                            modifier = Modifier.weight(1f, fill = false),
                        )
                    }
                    if (pending) {
                        if (category != null) Spacer(Modifier.width(8.dp))
                        Box(
                            modifier = Modifier
                                .clip(NeptuneTheme.shape.rFull)
                                .background(scheme.tertiaryContainer)
                                .padding(horizontal = 8.dp, vertical = 1.dp),
                        ) {
                            Text(
                                text = "pending",
                                style = typography.labelSmall.copy(
                                    color = scheme.onTertiaryContainer,
                                    fontWeight = FontWeight.W600,
                                ),
                            )
                        }
                    }
                }
            }
        }
        Spacer(Modifier.width(16.dp))
        // Trailing amount + time (bounded by the row, so they can never
        // overflow at narrow widths; the name column yields).
        Column(horizontalAlignment = Alignment.End) {
            Text(
                text = NeptuneTheme.formatDigits(amount),
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                style = money,
            )
            if (time != null) {
                Spacer(Modifier.height(2.dp))
                Text(NeptuneTheme.formatDigits(time), style = timeStyle)
            }
        }
    }
}

/**
 * A reward/voucher coupon: a big tabular value stub separated by a dashed
 * tear line from the title, an optional mono code chip, and an expiry
 * caption. Painted on `primaryContainer` at elevation-1.
 *
 * Web counterpart: `<npt-voucher-card>` · Flutter: `NeptuneVoucherCard`.
 * Theme-only, RTL-safe (the tear line sits on the inline-end edge of the
 * value stub).
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
public fun NeptuneVoucherCard(
    title: String,
    value: String,
    modifier: Modifier = Modifier,
    code: String? = null,
    expiry: String? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape
    val identity = NeptuneTheme.identity
    val display = rememberNeptuneFontFamily(NeptuneTheme.type.display)

    val onContainer = scheme.onPrimaryContainer
    val valueStyle = NeptuneTheme.moneyStyle(base = typography.headlineSmall)
        .copy(color = onContainer, fontWeight = FontWeight.W700)
    val codeStyle = NeptuneTheme.moneyStyle(base = typography.labelLarge).copy(
        color = scheme.onSurface,
        fontWeight = FontWeight.W600,
        letterSpacing = 1.2.sp,
    )
    val tear = scheme.outline

    Row(
        modifier = modifier
            // Elevation-1 at rest (web --npt-elev-1; the Dart 0/1/3 @ .20
            // shadow) — outside the clip.
            .nptShadow(identity.elevation1(scheme), shape.rLg)
            .clip(shape.rLg)
            .background(scheme.primaryContainer)
            .padding(20.dp)
            // Equal-height stub/body regardless of the parent's height (the
            // Dart IntrinsicHeight + stretch-Row recipe).
            .height(IntrinsicSize.Min),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // Value stub with a dashed tear line on the inline-end edge.
        Box(
            modifier = Modifier
                .fillMaxHeight()
                .drawBehind {
                    val stroke = 2.dp.toPx()
                    val x = if (layoutDirection == LayoutDirection.Rtl) {
                        stroke / 2
                    } else {
                        size.width - stroke / 2
                    }
                    drawLine(
                        color = tear,
                        start = Offset(x, 0f),
                        end = Offset(x, size.height),
                        strokeWidth = stroke,
                        pathEffect = PathEffect.dashPathEffect(
                            floatArrayOf(DASH_ON.dp.toPx(), DASH_OFF.dp.toPx()),
                        ),
                    )
                }
                .padding(end = 16.dp),
            contentAlignment = Alignment.Center,
        ) {
            Text(NeptuneTheme.formatDigits(value), style = valueStyle)
        }
        Spacer(Modifier.width(16.dp))
        // Body: title, code chip, expiry.
        Column(
            modifier = Modifier.weight(1f).fillMaxHeight(),
            verticalArrangement = Arrangement.Center,
        ) {
            Text(
                text = title,
                style = typography.titleMedium.copy(
                    fontFamily = display ?: typography.titleMedium.fontFamily,
                    fontWeight = FontWeight.W600,
                    color = onContainer,
                ),
            )
            if (code != null || expiry != null) {
                Spacer(Modifier.height(8.dp))
                FlowRow(
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    verticalArrangement = Arrangement.spacedBy(4.dp),
                ) {
                    if (code != null) {
                        Box(
                            modifier = Modifier
                                .align(Alignment.CenterVertically)
                                .clip(shape.rXs)
                                .background(scheme.surface)
                                .padding(horizontal = 12.dp, vertical = 4.dp),
                        ) {
                            Text(code, style = codeStyle)
                        }
                    }
                    if (expiry != null) {
                        Text(
                            text = expiry,
                            style = typography.bodySmall.copy(
                                color = onContainer.copy(alpha = 0.78f),
                            ),
                            modifier = Modifier.align(Alignment.CenterVertically),
                        )
                    }
                }
            }
        }
    }
}

/**
 * A scan-to-pay panel: a bordered QR frame sized [size], an optional
 * merchant caption, and a big tabular amount.
 *
 * Web counterpart: `<npt-qr-pay>` · Flutter: `NeptuneQrPay`.
 *
 * Real QR data is the consumer's job — fill the frame via [qr] (the web `qr`
 * slot); when null a placeholder QR mark renders at 70% of the frame.
 * Theme-only, RTL-safe.
 */
@Composable
public fun NeptuneQrPay(
    amount: String,
    modifier: Modifier = Modifier,
    merchant: String? = null,
    size: Dp = 180.dp,
    qr: (@Composable () -> Unit)? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val shape = NeptuneTheme.shape
    val money = NeptuneTheme.moneyStyle(base = typography.displaySmall)
        .copy(color = scheme.onSurface, fontWeight = FontWeight.W700)

    Column(
        modifier = modifier
            .clip(shape.rXl)
            .background(scheme.surfaceContainerLow)
            .padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        // The QR frame: 2dp outline-variant border on a surface pane.
        Box(
            modifier = Modifier
                .size(size)
                .clip(shape.rMd)
                .background(scheme.surface)
                .border(2.dp, scheme.outlineVariant, shape.rMd),
            contentAlignment = Alignment.Center,
        ) {
            CompositionLocalProvider(LocalContentColor provides scheme.onSurfaceVariant) {
                if (qr != null) {
                    qr()
                } else {
                    Icon(
                        NptWalletPayGlyphs.qrCode,
                        contentDescription = null,
                        modifier = Modifier.size(size * 0.7f),
                        tint = scheme.onSurfaceVariant,
                    )
                }
            }
        }
        if (merchant != null) {
            Spacer(Modifier.height(20.dp))
            Text(
                text = merchant,
                textAlign = TextAlign.Center,
                style = typography.bodyMedium.copy(color = scheme.onSurfaceVariant),
            )
        }
        Spacer(Modifier.height(8.dp))
        Text(NeptuneTheme.formatDigits(amount), style = money)
    }
}

/**
 * A selectable top-up option row: a squared icon tile slot, a label +
 * optional sublabel, a trailing tabular amount, and a chevron that mirrors
 * in RTL.
 *
 * Web counterpart: `<npt-topup-row>` · Flutter: `NeptuneTopupRow`.
 *
 * [icon] fills the 40dp `secondaryContainer` tile; when null the built-in
 * card-add glyph is used (the Dart default). Tappable via [onTap]
 * (≥56dp target). Theme-only, RTL-safe.
 */
@Composable
public fun NeptuneTopupRow(
    label: String,
    amount: String,
    modifier: Modifier = Modifier,
    sublabel: String? = null,
    icon: (@Composable () -> Unit)? = null,
    onTap: (() -> Unit)? = null,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val feedback = NeptuneTheme.feedback
    val haptics: HapticFeedback = LocalHapticFeedback.current
    val money = NeptuneTheme.moneyStyle(base = typography.bodyLarge)
        .copy(color = scheme.onSurface, fontWeight = FontWeight.W600)

    val clickableModifier = if (onTap != null) {
        Modifier.clickable {
            feedback.trigger(NptFeedbackCue.Tap, haptics)
            onTap()
        }
    } else {
        Modifier
    }

    Row(
        modifier = modifier
            .clip(NeptuneTheme.shape.rMd)
            .background(scheme.surfaceContainerLow)
            .then(clickableModifier)
            .defaultMinSize(minHeight = 56.dp)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // Icon tile.
        Box(
            modifier = Modifier
                .size(40.dp)
                .clip(NeptuneTheme.shape.rSm)
                .background(scheme.secondaryContainer),
            contentAlignment = Alignment.Center,
        ) {
            CompositionLocalProvider(LocalContentColor provides scheme.onSecondaryContainer) {
                if (icon != null) {
                    icon()
                } else {
                    Icon(
                        NptWalletPayGlyphs.cardAdd,
                        contentDescription = null,
                        modifier = Modifier.size(24.dp),
                    )
                }
            }
        }
        Spacer(Modifier.width(16.dp))
        // Label + sublabel (yield and ellipsize at narrow widths).
        Column(Modifier.weight(1f)) {
            Text(
                text = label,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                style = typography.bodyLarge.copy(color = scheme.onSurface),
            )
            if (sublabel != null) {
                Spacer(Modifier.height(2.dp))
                Text(
                    text = sublabel,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    style = typography.bodySmall.copy(color = scheme.onSurfaceVariant),
                )
            }
        }
        Spacer(Modifier.width(12.dp))
        Text(
            text = NeptuneTheme.formatDigits(amount),
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            style = money,
        )
        Spacer(Modifier.width(8.dp))
        Icon(
            NptWalletPayGlyphs.chevronEnd,
            contentDescription = null,
            modifier = Modifier.size(24.dp),
            tint = scheme.onSurfaceVariant,
        )
    }
}

/** The membership tier tones for [NeptuneTierBadge], mirroring the web
 * `tone` attribute (`gold`/`silver`/`primary`/`neutral`) and the Flutter
 * `NeptuneTierTone`. */
public enum class NeptuneTierTone { Neutral, Primary, Gold, Silver }

/**
 * A small membership tier pill: a leading dot + the tier label, coloured
 * from the theme by [tone] — primary/gold/silver take the matching container
 * roles; neutral sits on `surfaceContainerHighest` with an outline-variant
 * ring.
 *
 * Web counterpart: `<npt-tier-badge>` · Flutter: `NeptuneTierBadge`.
 * Theme-only, RTL-safe.
 */
@Composable
public fun NeptuneTierBadge(
    tier: String,
    modifier: Modifier = Modifier,
    tone: NeptuneTierTone = NeptuneTierTone.Neutral,
) {
    val scheme = MaterialTheme.colorScheme
    val typography = MaterialTheme.typography
    val display = rememberNeptuneFontFamily(NeptuneTheme.type.display)
    val pill = NeptuneTheme.shape.rFull

    val (bg, fg) = when (tone) {
        NeptuneTierTone.Primary -> scheme.primaryContainer to scheme.onPrimaryContainer
        NeptuneTierTone.Gold -> scheme.tertiaryContainer to scheme.onTertiaryContainer
        NeptuneTierTone.Silver -> scheme.secondaryContainer to scheme.onSecondaryContainer
        NeptuneTierTone.Neutral -> scheme.surfaceContainerHighest to scheme.onSurface
    }
    val ring = if (tone == NeptuneTierTone.Neutral) {
        Modifier.border(1.dp, scheme.outlineVariant, pill)
    } else {
        Modifier
    }

    Row(
        modifier = modifier
            .clip(pill)
            .background(bg)
            .then(ring)
            .defaultMinSize(minHeight = 24.dp)
            .padding(horizontal = 12.dp, vertical = 2.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            Modifier
                .size(6.dp)
                .clip(pill)
                .background(fg),
        )
        Spacer(Modifier.width(4.dp))
        Text(
            text = tier,
            style = typography.labelLarge.copy(
                fontFamily = display ?: typography.labelLarge.fontFamily,
                fontWeight = FontWeight.W700,
                color = fg,
                letterSpacing = 0.28.sp,
            ),
        )
    }
}
